text/0000-reseeding-rand.md

```diff
@@ -0,0 +1,657 @@
+it each time, it will be defined something like
+
+```
+struct ThreadRngInner {
+    inner: ChaChaRng,
+    generated_entropy: usize,
+}
+pub struct ThreadRng {
+    inner: Rc<RefCell<ThreadRngInner>>
+}
+```
+
+so that the internal state is shared between all instances of the
+random number generator.
+
+The `generated_entropy` field will record the number of bytes output
+by the RNG, and the whole `ThreadRng` will be reseeded from the
+operating system once this passes a certain threshold.
+
+#### Alternatives
+
+Do not regularly reseed from the operating system: only the OS RNG is
+designed for use in cryptographic circumstances and so the overhead
+may not be useful. Also, reseeding like this restricts the ability for
+a program to have forcibly deterministic execution by seeding the
+`ThreadRng`. Changing this is backwards compatible.
+
+This may make more sense to exist in `std::thread` or
+`std::thread_local`.
+
+### Alternatives (for "Provided RNGs")
+
+Repeating `Rng` in ever name may not be necessary, a scheme like the
+following may be interesting:
+
+- `Os`, `RandomDevice`, `Secure`, `SecureRng` (or just `OsRng`)
+- `ChaCha`
+- `Xorshift`
+- `std::rand::ThreadLocal` (a little confusing), `std::thread_local::Rng`
+
+We currently provide a `StdRng` type as an alias for another
+generator, this type disappears in this RFC, but could be reintroduced
+as a direct wrapper for `ChaChaRng`.
+
+## `Copy` and `Clone`
+
+Duplicating a random number generator can lead to generating
+correlated sequences of random numbers, which may not be a good idea.
+
+Random number generators should only implement `Copy` if copied values
+generate independent/uncorrelated sequences. Notably, this does not
+apply to PRNGs where the full internal state is stored inline, where
+copied values will have duplicated the state and hence all generate
+exactly the same sequence. This trait is particularly dangerous due to
+the implicit nature of the duplication.
+
+Random number generators can implement `Clone` in a way that
+duplicates the state and causes identical/correlated sequences. In
+fact, having identical sequences is likely one of the most valuable
+uses of this, to e.g. replay/re-record failing test runs. (D's
+`std.random` offers `save` on many RNGs for this purpose, and C++'s
+`<random>` has copy constructors implemented for most types.)
+
+### Alternatives
+
+Provide a non-generic way to duplicate an RNG, e.g. a manual `fn
+save(&self)` on each type. This gets annoying if the RNG has to be
+used in generic code in any way, although it is somewhat questionable
+if generic code duplicating an RNG is correct (that said, it could
+just be code that is generic over different RNG types, e.g. a general
+framework for rerunning/recording tests as mentioned above).
+
+## `std::rand::distributions`
+
+Moved from the standard library into a crates.io package.
+
+## Omissions
+
+The following is a (non-exhaustive) list of things which are explicitly not included in
```

Comment:
@huonw note, that I provided `SeedableRng` as an example (e.g.).

Some of removed functionality (for example `SeedableRng`) is either replaced by methods on structs (such as `ChaChaRng::from_seed()`) or has very minor use case (reseeding the Rng instead of creating a new one) for which I don’t see myself or anyone else depending on a crate(s). To me it sounds like such a crate would be doomed for one-digit (possibly infinitely close to zero) download count.